dev-docs/RFCs/v7.x/layer-culling-rfc.md

```diff
@@ -0,0 +1,94 @@
+# RFC: Layer Frustum Culling
+
+* Authors: Ib Green
+* Date: May, 2019
+* Status: **Draft with Open Issues**
+
+## Summary
+
+This RFC proposes adding the capability of layers to express bounding volumes, and to check those against the the current view (frustum) before GPU rendering, and avoid rendering layers that are not currently visible.
+
+
+## Overview
+
+We need to implement frustum culling for 3D tiles which is a substantial effort and this RFC proposes generalizing the implementation so that all deck.gl layers and use cases can benefit from this standard 3D engine feature.
+
+The ability to avoid rendering non-visible geometries is particularly important for large segmented data sets such as 3D Tile and potree point clouds, where hundreds of millions of points can be loaded into memory across hundreds of tiles.
+
+When moving in first-person view inside big point clouds, the idea is to only rendering things withing the current view frustum. This can lead to substantial performance increase, often 3-4x depending on direction of view and how tight the view frustum is, which can turn 5 FPS into 20 FPS.
+
+### math Support
+
+Culling requires intersecting bounding volumes with the view frustum, using a range of well known but not completely trivial 3D geometry techniques.
+
+In our case, bounding volumes (spheres and boxes) need to be fully 4x4 transformable (which means that the simple axis aligned bounding box classes used by many 3D libraries are not sufficient), but instead we need a more complex `OrientedBoundingBox`.
+
+The necessary math library (`BoundingSphere`, `OrientedBoundingBox`, `CullingVolume`, `PerspectiveFrustum` etc is being developed in the `@loaders.gl/3d-tiles` module, with the Cesium code base as reference.
+
+The current idea is to eventually move these classes out from loaders.gl and publish them as submodules of `math.gl`:
+
+```js
+import {OrientedBoundingBox, PerspectiveFrustum} from '@math.gl/culling';
```

Comment:
What is our criteria for adding new submodules? We have experienced a lot of issues in synchronizing dependency versions. If this is going to be a dependency of `@deck.gl/core`, might as well add it to math.gl proper?

Reply:
> What is our criteria for adding new submodules? 

Well, math.gl's front page promises a small bundle size for the Vector/Matrix classes. And in addition to these culling classes, for 3D tiles and similar formats we will need to add geospatial math classes, as well as geometry processing classes to our growing repertoire of math functions. 

I would certainly prefer to keep them outside of the math.gl core. As of today, these are in a new temporary submodule in loaders.gl. We will need to pick our poison. 

Whether culling should be a hard dependency of deck.gl's is a good question that this RFC raises. It is not necessary for classical visualizations, but we'll need the integration for certain cases. Perhaps we can make this pluggable.